File: onlinevote/src/main/java/com/example/vote/onlinevote/service/UserService.java
package com.example.vote.onlinevote.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.vote.onlinevote.model.User;
import com.example.vote.onlinevote.repository.UserRepository;

@Service
public class UserService {

    private final UserRepository userRepository;

    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public List<User> findNonCandidateUsers() {
        return userRepository.findByRoleNot("CANDIDATE");
    }
}

File: Kotlin/login1/src/main/kotlin/com/example/login1/service/UserService.kt
package com.example.login1.service

import com.example.login1.model.Users
import com.example.login1.repository.UserRepository
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserService(
    // Repository for database operations
    private val repo: UserRepository,

    // Spring Security authentication manager
    private val authManager: AuthenticationManager,

    // Service responsible for JWT generation
    private val jwtService: JWTService
) {

    // Password encoder (BCrypt with strength 12)
    private val encoder = BCryptPasswordEncoder(12)

    // Register a new user
    fun register(user: Users): Users {

        // Encode the raw password before saving
        val encodedPassword = encoder.encode(user.password)
            ?: throw IllegalStateException("Password encoding failed")

        // Create a new user object with the encoded password
        val userWithEncodedPassword = user.copy(
            password = encodedPassword
        )

        // Save the user to the database
        return repo.save(userWithEncodedPassword)
    }

    // Authenticate user credentials and return JWT if successful
    fun verify(users: Users): String {

        // Authenticate username and password
        val authentication: Authentication =
            authManager.authenticate(
                UsernamePasswordAuthenticationToken(
                    users.name,
                    users.password
                )
            )

        // If authentication is successful, generate JWT
        if (authentication.isAuthenticated) {
            return jwtService.generateToken(users.name)
        }

        // Authentication failed
        return "fail"
    }
}
